src/libs/safe-path/src/safe_dir_builder.rs

```diff
@@ -0,0 +1,162 @@
+        if !root.is_dir() {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", root.display()),
+            ));
+        }
+
+        let path = safe_join("/", path)?;
+        let mut suffix = path.strip_prefix(&root).map_err(|_| {
+            Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", path.display()),
+            )
+        })?;
+        if suffix.file_name().is_none() {
+            return SafePathBuf::from_path(root);
+        }
+        if !self.recursive {
+            if let Some(parent) = path.parent() {
+                root = root.join(parent);
+            }
+            // Safe to unwrap() because we have verified `suffix` is not empty.
+            suffix = Path::new(suffix.file_name().unwrap());
+        }
+
+        for comp in suffix {
+            let file = SafePathBuf::from_path(&root)?;
+            if !file.target().is_dir() {
+                return Err(Error::new(
+                    ErrorKind::Other,
+                    format!("Invalid path: {}", root.display()),
+                ));
+            }
+            root = root.join(comp);
+            DirBuilder::new()
+                .mode(self.mode)
+                .recursive(true)
+                .create(&root)?;
+        }
+
+        let result = SafePathBuf::from_path(&root)?;
+        if !result.target().is_dir() {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", root.display()),
+            ));
+        }
+
+        Ok(result)
+    }
+}
+
+impl Default for SafeDirBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::fs;
+
+    #[test]
+    fn test_safe_path_buf() {
+        let rootfs_dir = tempfile::tempdir().expect("failed to create tmpdir");
+        let rootfs_path = rootfs_dir.path();
+        let mut builder = SafeDirBuilder::new();
+
+        builder
+            .root(rootfs_path.join("__does_not_exist__"))
+            .unwrap_err();
+
+        fs::write(rootfs_path.join("txt"), "test").unwrap();
+        builder.root(rootfs_path.join("txt")).unwrap_err();
+
+        builder.root(rootfs_path).unwrap();
+        builder.create("/txt/a").unwrap_err();
+        let path = builder.create(rootfs_path.join(".")).unwrap();
+        assert_eq!(path.target(), rootfs_path);
+        builder.create(rootfs_path.join("a/b/c")).unwrap_err();
+        builder.create(rootfs_path.join("txt")).unwrap_err();
+
+        let path = builder.create(rootfs_path.join("a")).unwrap();
+        assert_eq!(path.target(), rootfs_path.join("a"));
+        assert!(rootfs_path.join("a").is_dir());
+
+        builder.recursive(true);
+        let path = builder.create(rootfs_path.join("a/b/c/d")).unwrap();
+        assert_eq!(path.target(), rootfs_path.join("a/b/c/d"));
+        assert!(rootfs_path.join("a/b/c/d").is_dir());
```

Comment:
Should you also check the permissions on the resulting directories (including the intermediate ones)?

Reply:
fixed